### PR TITLE
Adds ng-redux to the app.

### PR DESCRIPTION
**Pull request by SunburnedGoose**
This removes the need for scope in the controllers/services.  The redux pattern is how applications communicate between services and components in lieu of rootscope broadcasts and emits.

### CODE REVIEW

File: src/scripts/shell/dimAppCtrl.controller.js [hunk 1]
```diff
@@ -151,37 +129,33 @@ function DimApp(dimActivityTracker, dimState, ngDialog, $rootScope, loadingTrack
    * Show a popup dialog containing the given template. Its class
    * will be based on the name.
    */
-  function showPopupFunction(name) {
+  function showPopupFunction(name, translate) {
```
Comment (delphiactual) [inline src/scripts/shell/dimAppCtrl.controller.js]: remove `, translate`

File: src/scripts/shell/dimAppCtrl.controller.js [hunk 2]
```diff
@@ -151,37 +129,33 @@ function DimApp(dimActivityTracker, dimState, ngDialog, $rootScope, loadingTrack
...
     var result;
     return function(e) {
       e.stopPropagation();
 
+      var language = translate ? vm.settings.language + '/' : ''; // set language
```
Comment (delphiactual) [inline src/scripts/shell/dimAppCtrl.controller.js]: remove

File: src/scripts/shell/dimAppCtrl.controller.js [hunk 3]
```diff
@@ -151,37 +129,33 @@ function DimApp(dimActivityTracker, dimState, ngDialog, $rootScope, loadingTrack
...
+
       if (result) {
         result.close();
       } else {
         ngDialog.closeAll();
         result = ngDialog.open({
-          template: require('app/views/' + name + '.template.html'),
+          template: require('app/views/' + language + name + '.template.html'),
```
Comment (delphiactual) [inline src/scripts/shell/dimAppCtrl.controller.js]: remove `language + `

File: src/scripts/shell/dimAppCtrl.controller.js [hunk 4]
```diff
@@ -151,37 +129,33 @@ function DimApp(dimActivityTracker, dimState, ngDialog, $rootScope, loadingTrack
...
-        // ga('send', 'pageview', { page: '/' + name });
-        // }
       }
     };
   }
 
   vm.showSetting = showPopupFunction('settings');
-  vm.showAbout = showPopupFunction('about');
-  vm.showSupport = showPopupFunction('support');
+  vm.showAbout = showPopupFunction('about', true);
```
Comment (delphiactual) [inline src/scripts/shell/dimAppCtrl.controller.js]: remove `, true`

File: src/scripts/shell/dimAppCtrl.controller.js [hunk 5]
```diff
@@ -151,37 +129,33 @@ function DimApp(dimActivityTracker, dimState, ngDialog, $rootScope, loadingTrack
...
+  vm.showSupport = showPopupFunction('support', true);
```
Comment (delphiactual) [inline src/scripts/shell/dimAppCtrl.controller.js]: remove `, true`

**Comment by delphiactual** [comment]
looks like #1427 did not get merged into this

File: src/scripts/services/dimManifestService.factory.js [hunk 6]
```diff
@@ -1,250 +1,250 @@
...
+
+  return service;
+
+  /**
+   * Returns a promise for the manifest data as a Uint8Array. Will cache it on succcess.
+   */
+  function loadManifestRemote(version, language, path) {
+    service.statusText = $translate.instant('Manifest.Download') + '...';
+
+    return $http.get("https://www.bungie.net" + path + '?host=' + window.location.hostname, { responseType: "blob" })
```
Comment (bhollis) [inline src/scripts/services/dimManifestService.factory.js]: What's this for?

File: src/scripts/services/dimManifestService.factory.js [hunk 7]
```diff
@@ -1,250 +1,250 @@
...
-          }
-        });
-      }, function(error) {
-        reject(error);
-      });
-    });
-  }
-}
-
+/* global zip */
```
Comment (bhollis) [inline src/scripts/services/dimManifestService.factory.js]: I wonder why this file shows all lines changed?

File: src/scripts/shell/dimAppCtrl.controller.js [hunk 8]
```diff
@@ -61,14 +47,6 @@ function DimApp(dimActivityTracker, dimState, ngDialog, $rootScope, loadingTrack
   });
 
   hotkeys.add({
-    combo: ['r'],
```
Comment (bhollis) [inline src/scripts/shell/dimAppCtrl.controller.js]: Where does this hotkey get moved to?

File: src/scripts/services/dimManifestService.factory.js (hunk 6 above)
Comment (bhollis) [inline src/scripts/services/dimManifestService.factory.js]: (the hostname)

Comment (SunburnedGoose) [inline src/scripts/services/dimManifestService.factory.js]: Fixed an issue with chrome caching origin. You bump into it when you visit a website during testing, and then view it on beta site. Host name forces recache based on subdomain.

File: src/scripts/shell/dimAppCtrl.controller.js (hunk 8 above)
Comment (SunburnedGoose) [inline src/scripts/shell/dimAppCtrl.controller.js]: Refresh component I believe.

File: src/scripts/services/dimManifestService.factory.js (hunk 6 above)
Comment (bhollis) [inline src/scripts/services/dimManifestService.factory.js]: Cool!

File: src/scripts/services/dimManifestService.factory.js (hunk 7 above)
Comment (SunburnedGoose) [inline src/scripts/services/dimManifestService.factory.js]: Line endings. Don't know why yet. Guy is set autocrat to true.

**Comment by bhollis** [comment]
I'd love to take a look at this as just a diff from the router piece.

**Comment by delphiactual** [comment]
https://github.com/DestinyItemManager/DIM/compare/uirouter...ng-redux-stable

File: src/index.html [hunk 9]
```diff
@@ -27,13 +27,10 @@
...
-  <body ng-strict-di>
-    <div ng-controller="dimAppCtrl as app" ng-class="{ 'show-elements': app.settings.showElements, itemQuality: app.featureFlags.qualityEnabled && app.settings.itemQuality, 'show-new-items': app.settings.showNewItems, 'new-item-animated': app.settings.showNewAnimation }">
-      <div class="ng-cloak" ng-click="app.trackActivity()">
-        <div ui-view></div>
-      </div>
-    </div>
-
+  <body ng-app="app" ng-csp ng-strict-di ng-cloak>
+    <app>
+      Loading...
```
Comment (delphiactual) [inline src/index.html]: i18n?

Comment (SunburnedGoose) [inline src/index.html]: Angular hasn't loaded at this point so I don't have a means to apply i18n.  We can replace this with some animated graphic.  Spin DIM?  @chrisfried would finally have his baby see the light of day.

**Comment by bhollis** [comment]
@SunburnedGoose when you get a chance could you merge dev into this? I pushed a change to fix the line endings so the merge should be relatively smooth, but I'm not sure which changes to keep and which are just conflicts with the UI Router stuff.

**Comment by bhollis** [comment]
Looks like the "R" hotkey to refresh inventory doesn't work anymore - is there a way to rewire it?

The formatting changes in this PR obscure the scope of the change a bit, but I'm cautious - this is ~300 lines of new code, which is somewhat hard to follow (connecting, mapStateToThis, reducers, etc.), and four new dependencies. What practical benefit are we getting from this change? What made you want to make this change? Could this be done more incrementally, broken up into smaller changes? Where does this fit into the overall development plan of DIM? If we're changing the way we write code, what's the plan to make all the code work that way so we don't have a mishmash of different styles?

**Comment by SunburnedGoose** [comment]
Redux brings all the state into one service.  With all of our state in once location, we can rehydrate and dehydrate state when DIM starts up and shuts down.  You could easily prefill state from the server if that is data we have.

When we travel to a view, say our Loadout Builder, we can just present the last useful dataset that was generated so it loads quicker.  This is a great use case if none of the items changed light, or any new items appeared.  If there are new items, or an item has changed its light level, then we can dispatch an action to dump the state in those circumstances.  

State is immutable so any change will invoke digests (optimized in Angular 2+).  What's good about this is we can save state changes and allow for undos and rollbacks to an n-degree.  

Debugging gets interesting as you can see the actions moving across your state.  Some useful time-travelling debugging tools.

**Comment by SunburnedGoose** [comment]
And to go back to the OP, there is no more rootScope and pub/sub like we had with emit and broadcast.  You'll need an alternative to support the decoupling of services for state.  We've seen how large the list of injectables we need to get state, definitions, buckets, store, settings, etc...  Now it's just redux.
